### DISCUSSION TITLE
Is it possible to build a simple Reverse Proxy like Nginx with Mojo?

### DISCUSSION THREAD

**Discussion by s1037989** [mojolicious/mojo#2007] [Q&A] [answered]
IIUC, the example/proxy.pl works great as a forwarding proxy, allowing a Mojolicious app to behave like squid.

Is it possible to create something similar for a reverse proxy, allowing a Mojolicious app to behave like nginx?

I've looked at the example/proxy.pl and the Proxy section in the Cookbook, and I'm getting close in a lot of ways, but I feel like I'm missing something fundamental about what a reverse proxy does.  I've also tried reading all the various references to proxy throughout the Mojo docs, but I think the quantity of references is confusing me more than it's helping.

Does anyone know of any good resources for creating a reverse proxy or samples that I can look at to make a simple reverse proxy?

**Comment by kraih** [comment, marked as answer]
See https://docs.mojolicious.org/Mojolicious/Plugin/DefaultHelpers#proxy-get_p and https://docs.mojolicious.org/Mojolicious/Plugin/DefaultHelpers#proxy-start_p

**Reply by s1037989** [reply]
Thanks for the response!  I think I figured what is causing the proxy helpers to not work for me, and I wonder if I'm misusing it or if something else is amiss.

When the proxied server (4000, below) responds with an **empty body** (such as a 302 redirect), the proxy server mojo app (3000, below) eventually logs `Inactivity timeout` and at the same time the original requesting useragent closes with `Premature connection close`.  I have found that the proxied server (4000) does respond with the empty body and that the proxy server (3000)  does receive the response but does not pass it on to the originally requesting useragent.  If there was a single byte, the response would be forwarded and the requesting useragent would receive it.

I added a test to `t/mojolicious/proxy_app.t` to demonstrate the problem:

```
$ git diff t/mojolicious/proxy_app.t 
diff --git a/t/mojolicious/proxy_app.t b/t/mojolicious/proxy_app.t
index 1eadb6dbd..5f9474ade 100644
--- a/t/mojolicious/proxy_app.t
+++ b/t/mojolicious/proxy_app.t
@@ -16,6 +16,13 @@ my $daemon = Mojo::Server::Daemon->new(listen => ['http://127.0.0.1'], silent =>
 my $port   = $daemon->start->ports->[0];
 my $url    = Mojo::URL->new("http://127.0.0.1:$port");
 my $r      = $app->routes;
+$r->get(
+  '/empty' => sub {
+    my $c = shift;
+    $c->res->headers->header('X-Mojo-App' => 'Empty');
+    $c->render(data => '', status => 200);
+  }
+);
 $r->get(
   '/res1' => sub {
     my $c = shift;
@@ -102,6 +109,8 @@ get '/proxy3/:method/*target' => sub {
 my $t = Test::Mojo->new;
 
 subtest 'Various response variants' => sub {
+  $t->get_ok('/proxy1/empty')->status_is(200)->header_is('X-Mojo-App' => 'Empty')->content_is('');
   $t->get_ok('/proxy1/res1')->status_is(200)->header_is('X-Mojo-App' => 'One')->content_is('One!');
   $t->get_ok('/proxy1/res2')->status_is(200)->header_is('X-Mojo-App' => 'Two')->content_is('Two!');
   $t->get_ok('/proxy1/res3')->status_is(200)->header_is('X-Mojo-App' => 'Three')->header_is('X-Mojo-Method' => 'GET')
```

This also demonstrates the problem from the command line:

```
$ perl -Mojo -E 'a("/*x"=>{x=>""}=>sub{$_->proxy->get_p(l("http://localhost:4000")->path($_->param("x")))})->start' daemon -l http://*:3000 &
$ perl -Mojo -E 'a("/*y"=>{y=>""}=>sub{$_->render(text=>shift->param("y"))})->start' daemon -p -l http://*:4000 &
$ time mojo get 'http://localhost:3000/123qwe'
123qwe
real	0m0.255s
$ time mojo get 'http://localhost:3000/'
Premature connection close at /home/stefan/perl5/perlbrew/perls/perl-5.34.1/lib/site_perl/5.34.1/Mojolicious/Command/get.pm line 77.
real	0m30.260s
```

**Reply by kraih** [reply]
Please open an issue for the bug.

**Reply by s1037989** [reply]
Opened #2008 !

**Reply by s1037989** [reply]
I've been trying super hard to debug this issue -- using the Perl debugger and everything.  This is really hard.  Do you have any hints?  What module(s) I should be focusing on?  Breakpoints or watchers to use?

I cannot understand why a response code of 204 and content length 0 would be fine while 200 and 0 is not fine, but 200 and 1 is fine.  I can't see any code that would allow 204/0 but not 200/0.  It's extremely difficult to do comparisons between 200/0 and 200/1.  The only thing I've caught so far is that 200/0 never gets the HTTP response in the Mojo::IOLoop:Stream buffer, whereas 200/1 does get the HTTP response in the buffer.

With 200/0, [IO::Poll::_poll in Mojo::Reactor::Poll](https://github.com/mojolicious/mojo/blob/main/lib/Mojo/Reactor/Poll.pm#L55) times out after 5 seconds.

I am eager to keep diving into this, but I am as stuck as can be.  Any pointers would be very welcome.  Thank you!

**Reply by s1037989** [reply]
I posted a possibly reasonable patch (with tests) to the issue #2008.
